### PR TITLE
feat(flagging): mask rfi based on delay filtered visibilities

### PR DESCRIPTION
**Pull request by ssiegelx**
Adds three new tasks:
 - analysis.dayenu.DayenuDelayFilterFixedCutoff enables delay filtering of timeseries with rapidly changing frequency masks.
 - analysis.transform.ReduceChisq calculates a chi2-per-dof test statistic by comparing the delay filtered visibilities to the inverse noise variance stored in the weights.
 - analysis.flagging.RFIMaskChisqHighDelay generates a mask from this chi2-per-dof test statistic.

### CODE REVIEW

**Comment by ssiegelx** [comment]
Depends on #265 for the arPLS method.

File: draco/analysis/dayenu.py [hunk 1]
```diff
@@ -189,6 +188,147 @@ def _get_cut(self, prod):
...
+        weight = stream.weight[:].local_array
+
+        # Identify baselines that are always flagged
+        temp = np.any(weight > 0.0, axis=(0, 2))
+        baseline_flag = np.zeros_like(temp)
+        self.comm.Allreduce(temp, baseline_flag, op=MPI.LOR)
+        baseline_mask = ~baseline_flag
+
+        if np.all(baseline_mask):
+            self.log.error("All data flaged as bad.")
```
Comment (ljgray) [inline draco/analysis/dayenu.py:255]:
```suggestion
            self.log.error("All data flagged as bad.")
```

File: draco/analysis/dayenu.py [hunk 2]
```diff
@@ -189,6 +188,147 @@ def _get_cut(self, prod):
...
+            self.log.debug(f"Filter time {tt} of {ntime}. [{self.tauw:0.3f} microsec]")
+
+            # Construct the filter
+            try:
+                NF, index = highpass_delay_filter(
+                    freq, self.tauw, flag, epsilon=self.epsilon
+                )
+
+            except np.linalg.LinAlgError as exc:
+                self.log.error(f"Failed to converge while processing time {tt}: {exc}")
```
Comment (ljgray) [inline draco/analysis/dayenu.py:291]: I'm not too familiar with the behaviour of the dayenu filter - is it expected to throw `LinAlgError` every now and then? Or would it be better for the pipeline to fail here if there's an issue?

File: draco/analysis/flagging.py [hunk 3]
```diff
@@ -997,6 +997,253 @@ def process(
...
+        # For each frequency, caculate a weighted median
+        y = np.ascontiguousarray(y.astype(np.float64))
+        w = np.ascontiguousarray((~m).astype(np.float64))
+
+        med_y = weighted_median.weighted_median(y, w)
+        med_m = np.all(m, axis=-1)
+        med_w = np.ascontiguousarray((~med_m).astype(np.float64))
+
+        # Estimate a baseline
+        baseline = tools.arPLS_1d(med_y, mask=med_m, scale=self.reg_arpls)
```
Comment (ljgray) [inline draco/analysis/flagging.py:1173]:
```suggestion
        baseline = tools.arPLS_1d(med_y, mask=med_m, lam=self.reg_arpls)
```

This is on me, I changed this variable name in `arPLS_1d` to better reflect the standard variables used in the equation it's solving

**Review by ljgray — APPROVED** [review]
Looks good to me - just some minor comments

File: draco/analysis/dayenu.py (hunk 2 above)
Comment (ljgray) [reply draco/analysis/dayenu.py]: Sounds good

**Comment by ssiegelx** [comment]
I've confirmed that using the updated `DayenuDelayFilterFixedCutoff` task yields the same result as previously obtained using the three separate tasks.  If there are no objections I will go ahead and merge this PR and the corresponding one in ch_pipeline.

**Comment by ljgray** [comment]
> I've confirmed that using the updated `DayenuDelayFilterFixedCutoff` task yields the same result as previously obtained using the three separate tasks. If there are no objections I will go ahead and merge this PR and the corresponding one in ch_pipeline.

Sounds good to me

**Pull request merged**